File: src/main/java/de/fraunhofer/aisec/cpg/frontends/python/PythonLanguageFrontend.java
/*
 * Copyright (c) 2019, Fraunhofer AISEC. All rights reserved.
 *
 *  $$$$$$\  $$$$$$$\   $$$$$$\
 * $$  __$$\ $$  __$$\ $$  __$$\
 * $$ /  \__|$$ |  $$ |$$ /  \__|
 * $$ |      $$$$$$$  |$$ |$$$$\
 * $$ |      $$  ____/ $$ |\_$$ |
 * $$ |  $$\ $$ |      $$ |  $$ |
 * \$$$$$   |$$ |      \$$$$$   |
 *  \______/ \__|       \______/
 *
 */

package de.fraunhofer.aisec.cpg.frontends.python;

import de.fraunhofer.aisec.cpg.TranslationConfiguration;
import de.fraunhofer.aisec.cpg.frontends.LanguageFrontend;
import de.fraunhofer.aisec.cpg.frontends.TranslationException;
import de.fraunhofer.aisec.cpg.graph.TranslationUnitDeclaration;
import de.fraunhofer.aisec.cpg.passes.scopes.ScopeManager;
import de.fraunhofer.aisec.cpg.sarif.PhysicalLocation;
import io.github.oxisto.reticulated.ParserResult;
import io.github.oxisto.reticulated.PythonParser;
import io.github.oxisto.reticulated.ast.FileInput;
import io.github.oxisto.reticulated.ast.statement.Definition;
import io.github.oxisto.reticulated.ast.statement.Statement;
import java.io.File;
import org.checkerframework.checker.nullness.qual.Nullable;

/**
 * The language frontend for translating python language into the graph. It uses antlr to parse the
 * actual source code into an AST.
 */
public class PythonLanguageFrontend extends LanguageFrontend {

  private final StatementHandler statementHandler = new StatementHandler(this);
  private final StatementListHandler statementListHandler = new StatementListHandler(this);
  private final DefinitionHandler definitionHandler = new DefinitionHandler(this);
  private final SuiteHandler suiteHandler = new SuiteHandler(this);
  private final ExpressionHandler expressionHandler = new ExpressionHandler(this);
  private final SimpleStatementHandler simpleStatementHandler = new SimpleStatementHandler(this);

  public PythonLanguageFrontend(TranslationConfiguration config, ScopeManager scopeManager) {
    super(config, scopeManager, "");
  }

  @Override
  public TranslationUnitDeclaration parse(File file) throws TranslationException {
    PythonParser app = new PythonParser();
    ParserResult result = app.parse(file.getPath());
    FileInput input = result.getRoot();

    TranslationUnitDeclaration tu = new TranslationUnitDeclaration();

    for (Statement ctx : input.getStatements()) {
      // now things get a little tricky, since python does not distinguish between statements and
      // declarations, but the python parser has an utility class called 'Definition' to distinguish
      // class and function definitions from other statements such as 'if' and 'for'.
      if (ctx instanceof Definition) {
        tu.add(definitionHandler.handle((Definition) ctx));
      }

      // additionally, python allows statement on a global level, something we also do not allow, so
      // we need to put them into a virtual function

    }

    return tu;
  }

  @Override
  public <T> String getCodeFromRawNode(T astNode) {
    return null;
  }

  @Override
  public @Nullable <T> PhysicalLocation getLocationFromRawNode(T astNode) {
    return null;
  }

  @Override
  public <S, T> void setComment(S s, T ctx) {}

  public StatementHandler getStatementHandler() {
    return statementHandler;
  }

  public DefinitionHandler getDefinitionHandler() {
    return definitionHandler;
  }

  public SuiteHandler getSuiteHandler() {
    return suiteHandler;
  }

  public StatementListHandler getStatementListHandler() {
    return statementListHandler;
  }

  public ExpressionHandler getExpressionHandler() {
    return expressionHandler;
  }

  public SimpleStatementHandler getSimpleStatementHandler() {
    return simpleStatementHandler;
  }
}

File: src/main/java/de/fraunhofer/aisec/cpg/frontends/python/PythonLanguageFrontend.kt
package de.fraunhofer.aisec.cpg.frontends.python

import de.fraunhofer.aisec.cpg.ExperimentalPython
import de.fraunhofer.aisec.cpg.TranslationConfiguration
import de.fraunhofer.aisec.cpg.frontends.LanguageFrontend
import de.fraunhofer.aisec.cpg.frontends.TranslationException
import de.fraunhofer.aisec.cpg.graph.declarations.TranslationUnitDeclaration
import de.fraunhofer.aisec.cpg.passes.scopes.ScopeManager
import de.fraunhofer.aisec.cpg.sarif.PhysicalLocation
import java.io.File
import java.lang.Exception
import java.nio.file.Path
import jep.*

@ExperimentalPython
class PythonLanguageFrontend(config: TranslationConfiguration, scopeManager: ScopeManager?) :
    LanguageFrontend(config, scopeManager, ".") {
    companion object {
        @kotlin.jvm.JvmField var PY_EXTENSIONS: List<String> = listOf(".py")
    }

    @Throws(TranslationException::class)
    override fun parse(file: File): TranslationUnitDeclaration {
        return parseInternal(file.readText(Charsets.UTF_8), file.path)
    }

    override fun <T> getCodeFromRawNode(astNode: T): String? {
        // will be invoked by native function
        return null
    }

    override fun <T> getLocationFromRawNode(astNode: T): PhysicalLocation? {
        // will be invoked by native function
        return null
    }

    override fun <S, T> setComment(s: S, ctx: T) {
        // will be invoked by native function
    }

    private fun parseInternal(code: String, path: String): TranslationUnitDeclaration {
        val topLevel = Path.of("src/main/python")
        val entryScript = topLevel.resolve("main.py").toAbsolutePath()

        val tu: TranslationUnitDeclaration

        try {
            JepSingleton // configure Jep
            val interp = SubInterpreter(JepConfig().setRedirectOutputStreams(true))

            // TODO: extract main.py in a real python module with multiple files

            // load script
            interp.runScript(entryScript.toString())

            // run python function parse_code()
            tu = interp.invoke("parse_code", code, path, this) as TranslationUnitDeclaration
            interp.close()
        } catch (e: JepException) {
            e.printStackTrace()
            throw TranslationException("Python failed with message: $e")
        } catch (e: Exception) {
            throw e
        }

        return tu
    }
}
